include/swift/AST/ASTScope.h

```diff
@@ -10,117 +11,75 @@
+#include "swift/AST/DeclNodes.def"
+#undef DECL
+#undef ABSTRACT_DECL
+
+#define EXPR(Id, Parent) class Id##Expr;
+#include "swift/AST/ExprNodes.def"
+#undef EXPR
+
+#define STMT(Id, Parent) class Id##Stmt;
+#define ABSTRACT_STMT(Id, Parent) class Id##Stmt;
+#include "swift/AST/StmtNodes.def"
+#undef STMT
+#undef ABSTRACT_STMT
+
 class GenericParamList;
-class GuardStmt;
-class IfStmt;
-class IterableDeclContext;
-class LabeledConditionalStmt;
-class ParamDecl;
-class PatternBindingDecl;
-class RepeatWhileStmt;
-class SourceFile;
-class Stmt;
-class StmtConditionElement;
-class SwitchStmt;
-class TopLevelCodeDecl;
-class TypeDecl;
-class WhileStmt;
-
-/// Describes kind of scope that occurs within the AST.
-enum class ASTScopeKind : uint8_t {
-  /// A pre-expanded scope in which we know a priori the children.
-  ///
-  /// This is a convenience scope that has no direct bearing on the AST.
-  Preexpanded,
-  /// A source file, which is the root of a scope.
-  SourceFile,
-  /// The declaration of a type.
-  TypeDecl,
-  /// The generic parameters of an extension declaration.
-  ExtensionGenericParams,
-  /// The body of a type or extension thereof.
-  TypeOrExtensionBody,
-  /// The generic parameters of a declaration.
-  GenericParams,
-  /// A function/initializer/deinitializer.
-  AbstractFunctionDecl,
-  /// The parameters of a function/initializer/deinitializer.
-  AbstractFunctionParams,
-  /// The default argument for a parameter.
-  DefaultArgument,
-  /// The body of a function.
-  AbstractFunctionBody,
-  /// A specific pattern binding.
-  PatternBinding,
-  /// The scope introduced for an initializer of a pattern binding.
-  PatternInitializer,
-  /// The scope following a particular clause in a pattern binding declaration,
-  /// which is the outermost scope in which the variables introduced by that
-  /// clause will be visible.
-  AfterPatternBinding,
-  /// The scope introduced by a brace statement.
-  BraceStmt,
-  /// Node describing an "if" statement.
-  IfStmt,
-  /// The scope introduced by a conditional clause in an if/guard/while
-  /// statement.
-  ConditionalClause,
-  /// Node describing a "guard" statement.
-  GuardStmt,
-  /// Node describing a repeat...while statement.
-  RepeatWhileStmt,
-  /// Node describing a for-each statement.
-  ForEachStmt,
-  /// Describes the scope of the pattern of the for-each statement.
-  ForEachPattern,
-  /// Describes a do-catch statement.
-  DoCatchStmt,
-  /// Describes the a catch statement.
-  CatchStmt,
-  /// Describes a switch statement.
-  SwitchStmt,
-  /// Describes a 'case' statement.
-  CaseStmt,
-  /// Scope for the accessors of an abstract storage declaration.
-  Accessors,
-  /// Scope for a closure.
-  Closure,
-  /// Scope for top-level code.
-  TopLevelCode,
-};
+class TrailingWhereClause;
+class ParameterList;
+class PatternBindingEntry;
+class SpecializeAttr;
+class GenericContext;
+class DeclName;
+
+namespace ast_scope {
+class ASTScopeImpl;
+class GTXScope;
```

Comment:
The Doxygen comment for GTXScope says GenericType or Extension Scope. I know it's a new abbreviation to the code base -- I've seen NTD used a bunch -- and I generally agree with spelling things out. If I spell it out, we get:
`GenericTypeOrExtensionScope`
`GenericTypeOrExtensionWherePortion`
`GenericTypeOrExtensionWholePortion.`
Seems to be getting a bit long, but I'm sympathetic to the idea of spelling it out.

Reply:
I'll make sure that each class with GTX in its name includes a comment with an explanation of the abbreviation.

Reply:
Please spell it out. How often are you going to read vs write this code (also since you use Xcode, I imagine the IDE will spell it out for you via autocomplete). For dinosaurs like me who do not use a new IDE, not having it spelled out makes life difficult.

I need to look at the other two ones. There may be a better name.

Reply:
OK--spelled it out.